maturity-matrix.yaml

```diff
@@ -0,0 +1,161 @@
+# development time or software stability depending on their situation, so we
+# strive to separate self-reported maturity along these two axes.
+#
+# Needless to say, representations here are best-effort, and there is no
+# substitute for a mature release process. I.e., don't blindly deploy "mature"
+# OpenTelemetry components (or any software, for that matter!).
+productionMaturityLevels:
+  # "unknown" maturity is just what it purports to be. It's reasonable to
+  # assume "unstable", but we separate "unknown" and "unstable" to distinguish
+  # between a lack of evidence about production maturity and actual evidence of
+  # immaturity.
+  - unknown
+  # "unstable" components are not recommended for production workloads. They
+  # may crash the process, introduce performance artifacts, or have known
+  # security issues.
+  - unstable
+  # "beta" components have been used successfully in production workloads at
+  # scale, but either due to the uncertainty introduced by active development
+  # or a small number of production deployments, they should still be used
+  # cautiously in new deployments. They do not have known security issues.
+  - beta
+  # "stable" components are, to the best of the author's knowledge, safe for
+  # typical production use cases.
+  - stable
+
+# All OpenTelemetry APIs follow semver conventions (i.e., after v1.x,
+# backwards-incompatible changes should bump the major version number).
+apiMaturityLevels:
+  # "unimplemented" APIs do not exist yet for the component/language.
+  - unimplemented
+  # "alpha" maturity APIs can change in incompatible ways at any time.
+  - alpha
+  # "beta" maturity APIs should not introduce backwards-incompatible changes
+  # more than once every three months; and when those changes are introduced, the
+  # authors will make a best effort to provide compatibility bridges.
+  - beta
+  # "stable" maturity APIs should not introduce backwards-incompatible changes
+  # more than once every twelve months, and will make every effort to provide
+  # compatibility bridges if at all possible.
+  - stable
+
+
+###############################################################################
+# OPENTELEMETRY COMPONENTS
+#
+# Some OpenTelemetry components run within end-user applications and thus have
+# a footprint across many languages. Naturally, the production and API maturity
+# of these components varies from language to language. Other components run as
+# infrastructure in their own containers or VMs and thus have a single maturity
+# level regardless of the end-user's programming language.
+#
+# The maturity – or even the existence – of the "perLanguageComponents" varies
+# from language to language. These are components that integrate into the
+# actual end-user application runtime in production. See "languages" for the
+# actual maturity levels.
+perLanguageComponents:
+  - traceAPI
+  - metricsAPI
+  - contextAPI
+  - traceSDK
+  - metricsSDK
+  - contextSDK
+  - automaticInstrumentation
```

Comment:
should the integrations and exporters be also tracked? Zipking and Jaeger exporters, integrations like SpringBoot, etc.?

Reply:
IMO we should focus solely on SDK/API for trace/metric/context here and allow exporters and integrations to simply reference their stability or maturity against the lower level components.

Reply:
some integrations and exporters are the baseline for the stability as you cannot use API/SDK without ability to export it. Similar for integrations. I'm not sure how automaticInstrumentation here is different than integrations?

Reply:
Hm, that's a good point, although I don't know how to square it. Automatic instrumentation does seem to require integrations.

Reply:
+1, I think that end-users expect at least some of these (OT exporter, Jaeger exporter, Prometheus exporter, basic HTTP integrations) to be tracked

Reply:
I do _not_ want to suggest that we try to maintain a full compatibility matrix with framework-level things like Spring. The "registry" seems like the right place to track that sort of thing.

By "automaticInstrumentation", I meant specifically the idea of "no-code" (literally) installation of the OpenTelemetry instrumentation. This is an important requirement for many brownfield services that are no longer actively maintained but nevertheless participate in transactions. As such, I would not like to extend the definition to "1-line code changes" that nevertheless require recompilation.

@SergeyKanzhelev re your comment:

>some integrations and exporters are the baseline for the stability 

I disagree – they may be the baseline for "usefulness", but not for stability. 

Given the possibly large diversity of exporters in the future, I would personally prefer that we:

1. Adopt the API and production maturity levels for exporters, and
2. Track exporter maturity via the registry

Remember that we're not debating how this information is displayed on a website at the moment, we're just debating what should be considered "internal" to OpenTelemetry rather than a plugin/etc. The website can absolutely bring in the maturity levels for popular exporters in designated languages.

Reply:
We need to avoid situation when OT is only usable with the specific vendor exporter. This is why I think we should declare some baseline exporters here. It may be minimum functionality and not scalable, but must be reliable and stable.

Reply:
I wonder about the 'diversity of exporters' -- Yuri could probably weigh in here, but I suspect that many "vendors", both OSS and commercial, would like to simply accept OTel data format natively and thus not require an exporter at all.

With that in mind, this feels like a transient issue; Yes, there'll be maturity levels for exporters but over time those exporters simply won't matter since your backend will accept the data with no translation.

Reply:
> I suspect that many "vendors", both OSS and commercial, would like to simply accept OTel data format natively and thus not require an exporter at all.

Not immediately. It may take a while

> With that in mind, this feels like a transient issue; Yes, there'll be maturity levels for exporters but over time those exporters simply won't matter since your backend will accept the data with no translation.

Yes, but if we intent to track progress towards 1.0 I'd suggest to include tracking of this transient issue

Reply:
@SergeyKanzhelev I'm going to continue to push back on this... if I want to use Jaeger's exporter and it's the only thing that's mature for my given language, IMO that's fine. Now, whether the SIGs would feel comfortable marking their APIs as "stable" (or even "beta") with zero or one exporter implemented is a different question.